test/haxesharp/collections/LinqTest.hx

```diff
@@ -50,6 +51,55 @@ class LinqTest
     }
 
     @Test
+    public function singleThrowsIfArrayIsNull()
+    {
+        Assert.throws(ArgumentNullException, (_) => Linq.single(null));
+        Assert.throws(ArgumentNullException, (_) => Linq.single(null, (x) => true));
+    }
+
+    @Test
+    public function singleWithoutPredicateThrowsIfArrayIsEmpty()
+    {
+        Assert.throws(InvalidOperationException, (_) => [].single());
+    }
+
+    @Test
+    public function singleWithoutPredicateThrowsIfArrayContainsMultipleElements()
+    {
+        Assert.throws(InvalidOperationException, (_) => [1,2,3].single());
+    }
+
+    @Test
+    public function singleWithoutPredicateReturnsTheElementInASingleElementArray()
+    {
+        Assert.that([1].single(), Is.equalTo(1));
+    }
+
+    @Test
+    public function singleWithPredicateThrowsIfArrayIsEmpty()
+    {
+        Assert.throws(InvalidOperationException, (_) => [].single((x) => true));
+    }
+
+    @Test
+    public function singleWithPredicateThrowsIfNoElementSatisfiesPredicate()
+    {
+        Assert.throws(InvalidOperationException, (_) => [1,2,3].single((x) => x == 4));
+    }
+
+    @Test
+    public function singleWithPredicateThrowsIfMultipleElementsSatisfyPredicate()
+    {
+        Assert.throws(InvalidOperationException, (_) => [1,2,3,2].single((x) => x == 2));
+    }
+
+    @Test
+    public function singleWithPredicateReturnsTheElementIfOnlyOneElementSatisfiesPredicate()
+    {
+        Assert.that([1,2,3].single((x) => x == 3), Is.equalTo(3));
+    }
+
+    @Test

```

Comment:
Your test coverage is excellent, well done. I would recommend adding a case to trigger when the array is null (that check I commented on earlier).

Reply:
The think the test `singleThrowsIfArrayIsNull` I added covers this? I could split it into two tests, one for single with predicate and one for single without predicate if that seems cleaner.

Reply:
Ah, sorry, I didn't notice that. That's fine.